drivers: eeprom: EEPROM emulation in flash memory

Laczen opened the pull request:
This driver emulates a EEPROM device in flash.

**Updated description to reflect changes in eeprom representation in flash.**

The emulation represents the EEPROM in flash as a region that is a direct map of the eeprom data followed by a region where changes to the eeprom data is stored. Changes are written as address-data combinations. The size
of such a combination is determined by the flash write block size and the size of the eeprom (required address space), with a minimum of 4 byte.
The eeprom page needs to be a multiple of the flash page. Multiple eeprom pages is also so supported and increases the number of writes that can be performed.

The eeprom size, pagesize and the flash partition used for the eeprom are defined in the dts. The flash partition should allow at least two eeprom pages.

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

Laczen committed:
drivers: eeprom: EEPROM emulation in flash memory

This driver emulates a EEPROM device in flash.

Reworked implementation with modified flash layout.

The emulation represents the EEPROM in flash as a region that is a
direct map of the eeprom data followed by a region where changes to
the eeprom data is stored. Changes are written as address-data
combinations. The size of such a combination is determined by the
flash write block size and the size of the eeprom (required address
space), with a minimum of 4 byte.
The eeprom page needs to be a multiple of the flash page. Multiple
eeprom pages is also so supported and increases the number of writes
that can be performed.

The eeprom size, pagesize and the flash partition used for the eeprom
are defined in the dts. The flash partition should allow at least two
eeprom pages. For fast read access a rambuffer can be enabled for the
eeprom (by setting the option rambuf in the dts).

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

henrikbrixandersen commented:
@Laczen Thank you for the PR. Could you please elaborate a bit on the purpose of emulating an EEPROM in flash? What is the motivation/use-case behind this functionality?

Laczen commented:
> @Laczen Thank you for the PR. Could you please elaborate a bit on the purpose of emulating an EEPROM in flash? What is the motivation/use-case behind this functionality?

The motivation is to provide a eeprom like device but without an eeprom being added physically. Several devices provide such a interface (I think the stm32 eeprom is really a software emulation). The current PR is a generalized approach that provides an eeprom like device for all zephyr devices that have flash.

The main advantage of an eeprom device is that it can be rewritten multiple times, the PR does the same, but on a flash device without requiring a erase when a rewrite to an address is done. The proposed solution would not reach the same amount of writes as an eeprom, but a lot more than using plain flash writes.

A typical use for this interface would be to store certain values using a address instead of a name (the same way an eeprom is used).

I hope this answers your question.

nvlsianpu reviewed (commented):
Idea looks fine to me, I remember something like that was subject of one great Energy Micro application note (part of silabs today). Will review.
I wonder why the `eeprom page size` must be bigger than the twice of `the eeprom size` and not equal or bigger.

Laczen commented:
> I wonder why the `eeprom page size` must be bigger than the twice of `the eeprom size` and not equal or bigger.

Because when `2*size == page_size` and the eeprom is fully used also the page is fully used, you cannot perform a copy and then do a write as your page is full again. It would be possible to avoid this but then you would get a page rewrite for every write, it is simpler and better to make the eeprom size smaller than half the page size and leave some space for writing changes.

Laczen commented:
Maybe the storage method (data + address + CRC16) is not the best. I am considering changing the storage as a direct storage on flash of the eeprom area followed by a set of changes.

To compare the 2 methods (1536 byte eeprom on 4kB page):
a. Present method:
    a. when eeprom is fully used for 1536 byte: space for 1024 byte (512 eeprom byte) of changes per page
    b. when eeprom is only used for 1024 byte: space for 2048 byte (1024 eeprom byte) of changes per page
    c. when eeprom is only used for 768 byte: space for 2560 byte (1280 eeprom byte) of changes per page
b. Alternative method: regardless of how much eeprom is utilized there is always 2560 byte (1280 eeprom byte) for changes.

Laczen commented:
The driver has been updated to the new idea for storage. It has also been updated to use `DEVICE_DT_INST_DEFINE()`.

nvlsianpu reviewed (commented):
I think `the eeprom simulator` is still nice feature of qemu_x86 which can be used for verify storages which depends on EEPROM.
'The eeprom emulator` has another purpose - it can be used in really devices as a storage.

Would be great to have them both and test them both.

Laczen commented:
> I think `the eeprom simulator` is still nice feature of qemu_x86 which can be used for verify storages which depends on EEPROM.
> 'The eeprom emulator` has another purpose - it can be used in really devices as a storage.
> 
> Would be great to have them both and test them both.

@nvlsianpu, thanks for the comment. I agree that the eeprom simulator has a different purpose than the eeprom emulator. As proposed the eeprom simulator is indeed no longer tested on qemu_x86, but is is still tested as part of native_posix. On qemu_x86 the emulator can replace the simulator as they are both ram representations of either eeprom or flash.
I am trying to add both the simulator and the emulator to qemu_x86 but I am getting devicetree errors when defining both in the dts. I don't know if it is even possible to define two different compatibles in one dts file.

Laczen commented on drivers/eeprom/eeprom_simulator.c:
```diff
@@ -98,7 +98,7 @@ static int eeprom_range_is_valid(const struct device *dev, off_t offset,
 	const struct eeprom_sim_config *config = DEV_CONFIG(dev);
 
-	if ((offset + len) < config->size) {
+	if ((offset + len) <= config->size) {
```
True, this is a bugfix to existing feature, however it only pops up when doing the more extensive testing that is also provided with the present PR. Doing it separately would mean that the present PR would be failing until the separate PR is approved. This is why I included it in the PR (which is both a new EEPROM and more extensive testing).

sjg20 replied:
A separate commit in the same PR seems fine to me, but it should not be in this commit

Laczen commented:
> > I think `the eeprom simulator` is still nice feature of qemu_x86 which can be used for verify storages which depends on EEPROM.
> > 'The eeprom emulator` has another purpose - it can be used in really devices as a storage.
> > Would be great to have them both and test them both.
> 
> @nvlsianpu, thanks for the comment. I agree that the eeprom simulator has a different purpose than the eeprom emulator. As proposed the eeprom simulator is indeed no longer tested on qemu_x86, but is is still tested as part of native_posix. On qemu_x86 the emulator can replace the simulator as they are both ram representations of either eeprom or flash.
> I am trying to add both the simulator and the emulator to qemu_x86 but I am getting devicetree errors when defining both in the dts. I don't know if it is even possible to define two different compatibles in one dts file.

Updated qemu_x86.dts to include both simulator and emulator eeprom.

sjg20 commented on drivers/eeprom/Kconfig.eeprom_emu:
```diff
@@ -0,0 +1,13 @@
+	bool "Emulated EEPROM driver"
+	select FLASH
+	help
+	  Enable emulated (on flash) EEPROM support.
```
This needs a lot more detail. See your commit message for ideas

Laczen replied:
done

sjg20 commented on dts/bindings/mtd/zephyr,emu-eeprom.yaml:
```diff
@@ -0,0 +1,18 @@
+    pagesize:
+      type: int
+      required: true
+    partition:
```
These need descriptions

Laczen replied:
done

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+	address = 0;
+	while (address < size) {
+		rc = eeprom_write(eeprom, address, wr_buf2, sizeof(wr_buf2));
+		zassert_equal(0, rc, "Unexpected error code (%d)", rc);
```
This seems to be writing the whole EEPROM. If so, why is that necessary for a test? It should be possible to test things with just a few writes.

Laczen replied:
done

sjg20 replied:
Actually it is, because it does not increase test coverage, but adds more time. It is harder to debug a failing test that does way more than it needs to. The log would be enormouse!

Tests should be written to cover the functionality present.

Laczen replied:
Ok, I will reduce it.

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+	}
+}
+
+static void test_write_and_verify_var_addr(void)
```
What is a var address?

I think this needs a comment

Laczen replied:
done

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+	eeprom = device_get_binding(eeprom_label);
+
+	address = 0;
+	while (address < 4) {
```
for loop?

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+	eeprom = device_get_binding(eeprom_label);
+
+	address = 0;
+	while (address < 16) {
```
for loop?

Laczen replied:
done

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+static void test_write_and_verify_byte(void)
+{
+	const uint8_t wr_buf1[1] = { 0x00 };
+	uint8_t rd_buf[sizeof(wr_buf1)];
```
Since read and write size are the same, how about

const size = 1;

and then use that instead of sizeof()?

Laczen replied:
done

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+
+	address = 0;
+	for (int cnt = 0; cnt < 16; cnt++) {
+
```
drop blank line, and at 133

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+	size = eeprom_get_size(eeprom);
+
+	address = 0;
+	for (int cnt = 0; cnt < 16; cnt++) {
```
s/cnt/i/ It doesn't seems like it is counting anything

Laczen replied:
done

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+	eeprom = device_get_binding(eeprom_label);
+	size = eeprom_get_size(eeprom);
+
+	address = 0;
```
const offf_t address = 0

Laczen replied:
done

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+static void test_constant_address(void)
+{
+	const uint8_t wr_buf1[4] = { 0xFF, 0xEE, 0xDD, 0xCC };
+	uint8_t rd_buf[sizeof(wr_buf1)];
```
const size = ...

Laczen replied:
This is an inheritance from the original test

sjg20 commented on tests/drivers/eeprom/src/main.c:
```diff
@@ -43,27 +43,148 @@ static void test_write_and_verify(void)
+
+static void test_constant_address(void)
+{
+	const uint8_t wr_buf1[4] = { 0xFF, 0xEE, 0xDD, 0xCC };
```
Does Zephyr use lower-case hex? I find it much easier to read

Laczen replied:
This is an inheritance from the original test.

Laczen commented:
@sjg20, thanks for reviewing the PR. I have updated the code to correct your remarks.

sjg20 commented on drivers/eeprom/Kconfig.eeprom_emu:
```diff
@@ -0,0 +1,16 @@
+	  Enable emulated (on flash) EEPROM support. This mimics an external
+	  EEPROM on a flash partition. The number of writes than can be
+	  performed to the EEPROM is maximized by using a flash area that is
+	  larger the EEPROM area and by storing only changes to the EEPROM data.
```
larger than?

Laczen replied:
Yes, and also 'The number of writes tha**t**...'

sjg20 reviewed (changes requested):
I don't like github reviews very much.

Can I just check that you have two separate commit messages? On github it looks like they are the same.

Laczen commented:
> I don't like github reviews very much.
> 
> Can I just check that you have two separate commit messages? On github it looks like they are the same.

if you hit the ... next to the commit title it shows the complete message, they are different.

sjg20 commented:
> > I don't like github reviews very much.
> > Can I just check that you have two separate commit messages? On github it looks like they are the same.
> 
> if you hit the ... next to the commit title it shows the complete message, they are different.

OK I see, thanks. That is so confusing to show the PR title as the title of each commit.

henrikbrixandersen commented:
> The motivation is to provide a eeprom like device but without an eeprom being added physically. Several devices provide such a interface (I think the stm32 eeprom is really a software emulation). The current PR is a generalized approach that provides an eeprom like device for all zephyr devices that have flash.

I believe another reason for using built-in device EEPROM emulation is speed. The internal EEPROMs are normally RAM-backed with internal background synchronisation to flash. The implementation presented here does not provide such a speed benefit over using flash with a file system.

> The main advantage of an eeprom device is that it can be rewritten multiple times, the PR does the same, but on a flash device without requiring a erase when a rewrite to an address is done. The proposed solution would not reach the same amount of writes as an eeprom, but a lot more than using plain flash writes.

How so? Using flash as a backend will only give you the maximum write count of flash - no matter how you distribute the writes.

> A typical use for this interface would be to store certain values using a address instead of a name (the same way an eeprom is used).

Have you considered using a filesystem instead?

Laczen commented:
@henrikbrixandersen, thanks for the feedback.

> I believe another reason for using built-in device EEPROM emulation is speed. The internal EEPROMs are normally RAM-backed with internal background synchronisation to flash. The implementation presented here does not provide such a speed benefit over using flash with a file system.

It is correct that most internal EEPROMs are RAM-backed. If however a device is used that has only limited RAM, you do not wish to use this for the EEPROM emulation. I don't know why in a internal emulation you are forced to use RAM, we know EEPROM is generally not fast, so why should an emulated solution be fast.

If you would like to have this speed increase it is straightforward to add a RAM buffer for the EEPROM, provide a fast read from this buffer and call eeprom-write routines whenever something is changed to the RAM buffer. This code (buffered-EEPROM) could then be used by external EEPROM as well.
 
> 
> > The main advantage of an eeprom device is that it can be rewritten multiple times, the PR does the same, but on a flash device without requiring a erase when a rewrite to an address is done. The proposed solution would not reach the same amount of writes as an eeprom, but a lot more than using plain flash writes.
> 
> How so? Using flash as a backend will only give you the maximum write count of flash - no matter how you distribute the writes.
>

The limiting factor on flash devices is the number of (page) erases, not the number of writes. By distributing the writes, page erases are avoided and therefore it is possible to do a lot more writes.

> > A typical use for this interface would be to store certain values using a address instead of a name (the same way an eeprom is used).
> 
> Have you considered using a filesystem instead?

This is the typical usage for eeprom, I don't think we need to question this. But what are the advantages over using a filesystem:
* The emulation code size is smaller, leaving more space to emulate the eeprom.
* Writing changes to the eeprom requires less space than a write to a filesystem (e.g. in the eeprom the minimal change is 4 bytes (for a 2 byte change), on nvs this same change will require at least: 4 byte (data) + 8 byte (filesystem overhead).

That there are device specific internal implementations for such a eeprom emulation should provide sufficient reason to implement this solution, it wouldn't be there if nobody needs it.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		rc = flash_write_protection_set(dev_config->flash_dev, false);
```
This is rather critical error that will make write impossible, yet there is no error reported, nor code returned to inform a caller that the sub-routine failed.

Laczen replied:
I will turn around the code a little bit, check if the device is available at start of eeprom read/write, disable the write protection if required and make the flash routines inlines.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+	memset(buf, 0x00, sizeof(buf));
+
+	offset += (dev_config->page_size - sizeof(buf));
+	eeprom_emu_flash_write(dev, offset, buf, sizeof(buf));
```
The `eeprom_emu_flash_write` neither reports error nor returns error in case of failure which means that the entire  `eeprom_emu_page_invalidate` may be unsuccessful with not information about that fact.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		rc = flash_write_protection_set(dev_config->flash_dev, false);
```
This is rather critical error that does not allow to complete further actions successfully.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+	}
+
+	/* erase the new page */
+	eeprom_emu_flash_erase(dev, next_page_offset, dev_config->page_size);
+
+	/* copy existing data */
+	while (rd_offset < dev_config->size) {
+		eeprom_emu_word_read(dev, rd_offset, buf);
+		if (eeprom_emu_is_word_used(dev, buf)) {
+			eeprom_emu_flash_write(dev, next_page_offset +
+					       rd_offset, buf, sizeof(buf));
+		}
+		rd_offset += sizeof(buf);
+	}
+	/* invalidate the old page */
+	eeprom_emu_page_invalidate(dev, dev_data->page_offset);
+	dev_data->write_offset = dev_config->size;
+	dev_data->page_offset = next_page_offset;
```
The `eeprom_emu_flash_erase`, the `eeprom_emu_flash_write` and `eeprom_emu_page_invalidate` may all fail here due to `flash_write_protection_set` yet the entire code will proceed , which will basically mean that compaction may not happen at all, yet the `dev_data` will be updated.

henrikbrixandersen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+	if (!device_is_ready(dev_config->flash_dev)) {
+		__ASSERT(0, "Could not get flash device binding");
+	}
+
```
Since this is a software-only implementation of an EEPROM-like flash filesystem, I think we should add a version field to the stored data. Otherwise, we will never be able to change the flash representation.

Laczen replied:
At first I was thinking, yes I can add this to the last record that gets overwritten with all 0. But then, the configuration is all done by the dts, there is no software configuration of the flash representation. Making changes to the dts is like making hardware changes, no compatibility guaranteed. That said, would it be possible to convert from one eeprom_emulation to a different one: yes, create two devices (different version) using the same partition: read in to a ram buffer using the old representation, erase the partition and write the ram buffer using the new representation. This method could even be used when making changes to the dts file.

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+ * of such a combination is determined by the flash write block size and
+ * the size of the eeprom (required address space), with a minimum of 4 byte.
+ * The eeprom page needs to be a multiple of the flash page. Multiple eeprom
+ * pages is also so supported and increases the number of writes that can be
```
Yes regarding the comment (have changed this).
Regarding how this is done: the EEPROM is represented in flash as an area of size EEPROM and an area for changes. Whenever a write needs to be performed the library checks if it can do a direct write to the first area (and does so if still empty), if not it stores a change in the second area (address + new data). If there is no more space to store changes a new page is taken, this first writes the first area with the latest values and then starts writing changes.
The ratio between the EEPROM page size and EEPROM size determines the amount of changes that can be written before a page is erased. A EEPROM that would be used to mainly store constant data and only limited changes would use a smaller ratio than a EEPROM that would be used to store frequently changing data. A ratio of 2 between EEPROM page size and EEPROM size (and a partition of 2 EEPROM pages) guarantees that each of the EEPROM cells can be written at least 0.5 * 2 * number of page erases (this is only considering the changes). This is a lot more than in case a write too a next page is performed, this would only allow (number of page erases)/(EEPROM size) changes.

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+ * pages is also so supported and increases the number of writes that can be
+ * performed.
+ *
+ * Internally the address of an eeprom byte is represented by a uint32_t (this
+ * should be sufficient in all cases). In case the eeprom size is smaller than
+ * 64kB only a uint16_t is used to store changes. In this case the change stored
```
The first sentence is talking about the address internally (uint32_t), the second is about how changes are stored. I will try to make this clearer.

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	int rc;
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		rc = flash_write_protection_set(dev_config->flash_dev, false);
+		rc |= flash_write(dev_config->flash_dev, offset, blk, len);
+		rc |= flash_write_protection_set(dev_config->flash_dev, true);
+		if (!rc) {
+			return;
+		}
+	}
+	__ASSERT(0, "Flash interface failed");
```
Because it should also assert if there is a flash error. 
I am trying to rework the flash write/read/erase routines to allow better error return codes, this however is not so easy because the eeprom_api returns `int` instead of `ssize_t`. And an api change is not on my to do list. An idea ?

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if (!dev) {
```
This is to catch user errors that would call eeprom_size(NULL), the same for eeprom_write(), eeprom_read(). It is allowed (although not very usefull) to create a eeprom of zero size.  The return code could also have been `-EINVAL`, but it was already `0` in eeprom_simulator.c.

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,543 @@
+
+#define EEPROM_PARTITION(n) DT_PHANDLE_BY_IDX(DT_DRV_INST(n), partition, 0)
+
+#define GPARENT(node_id) DT_PARENT(DT_PARENT(node_id))
```
Yes, but this is a very recent addition.

Laczen commented:
@de-nordic, thanks for the review. I have updated the code to correct your remarks. Regarding the documentation of the dts parameters, this is done in the yaml description items. All other I think are covered.

carlescufi commented:
@Laczen if I understand correctly your objective with this PR is to present something akin to a lightweight filesystem that is backed by flash memory and exposes itself through the EEPROM API. My first question is therefore, wouldn't it be better to create a new API+layer, free of the EEPROM term and restrictions, that provides all the functionality you need without having to "shoehorn" an existing API for a class of devices (EEPROMs) on top of another (flash memories)? Something like "lnvm" (linear non-volatile memory) perhaps? The API could be very similar to the existing EEPROM one, but it would be a brand new one.

Laczen commented:
@carlescufi, thank you for the comment.
The idea is not to present something akin to a lightweight filesystem. The idea is really to emulate an EEPROM: an area that you can address directly and do frequent writes to. This emulation is done by several manufacturers to avoid the necessity to add a physical EEPROM. You can find application notes from e.g. ST describing such a EEPROM emulation layer. Also arduino has something similar on devices without EEPROM (e.g. esp8266).
The PR presents a representation of the EEPROM in flash as a region that is a direct map to the EEPROM area followed by a region that stores changes written to the EEPROM. A different approach could have been to store this representation in a filesystem (this is e.g. used on arduino for esp32), I don't know if this would be possible in zephyr, but it's not the way that I have chosen.
The functionality aimed by the PR is to provide an EEPROM like device without adding an EEPROM, nothing more, nothing less. I don't see any need for creating a different kind of flash device.

de-nordic commented:
@carlescufi , @Laczen 
[ST EEPROM Emulation](https://www.st.com/resource/en/application_note/dm00049914-eeprom-emulation-in-stm32f0xx-microcontrollers-stmicroelectronics.pdf)

I can see two cases for this emulation
1) You are phasing out the EEPROM from your hardware (because $$$)
So you get your prototype board that no longer has EEPROM yet somehow there are more important things to do now than to switch to flash: this emulation basically gives you some time till you are able to flush your backlog.
2) You are going to use EEPROM but it is not there yet
For example you are working with your DK and there is no possible way, for some reason, to connect EEPROM to it right now, because of PCB design and GPIO occupancy for example. The prototype PCB is still in design and will be available in several weeks. This layer will let you code like you have that EEPROM, which will be "DTS lie" , and then modify your DTS to switch to the real one when the prototype is available.

The 3rd case would be long time system support, where you have to switch to new hardware while software will be only recompiled with no changes to logic, features, etc.

carlescufi commented:
@Laczen and @de-nordic thanks for the respective comments and insight.

>This emulation is done by several manufacturers to avoid the necessity to add a physical EEPROM.

Thanks for the link to the STM emulation layer. Could you provide a couple more links to other manufacturers' emulation layers in order for us to properly document in this discussion that this is indeed common practice in the industry? This would help because it would provide us with a strong argument in favor of reusing the EEPROM API instead of defining a new one. Thanks!

carlescufi reviewed (approved):
The approach makes sense to me and in general the implementation LGTM.

pabigot commented on dts/bindings/mtd/zephyr,emu-eeprom.yaml:
```diff
@@ -0,0 +1,21 @@
+properties:
+    size:
+      required: true
+      description: EEPROM size (byte)
```
IMO this shouldn't have a `description` since this is just changing the `required` field of an existing `size` property that is documented in `eeprom-base.yaml`.

pabigot reviewed (commented):
I have no informed opinion on whether this is useful.  Also I'm not going to nit-pick the style issues, except to note that this is inconsistent about whether the finale statement block of `if`/`while`/etc control structures is followed by a blank line.  So I plan to abstain relative to whether this should be accepted.

One retracted venting related to ongoing Zephyr mis-use of signed types where unsigned ones are appropriate, change not expected; one change requested for the binding YAML; and before merge I want the commits squashed, since the second simply corrects gaps in the first.  (That one I *would* block on.)

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,634 @@
+
+struct eeprom_emu_data {
+	/* Offset in current (EEPROM) page where next change is written */
+	off_t write_offset;
```
Yes we did discuss this, the way I look at it now: all the cases where `off_t` are used are considered as relative addresses, these can be positive or negative, so I'm not stuck anymore. The main problem I'm still having with `off_t` is that it needs to be cast for logging/printing.

Laczen commented:
Hi @pabigot, thanks for the comments.

> I have no informed opinion on whether this is useful. Also I'm not going to nit-pick the style issues, except to note that this is inconsistent about whether the finale statement block of `if`/`while`/etc control structures is followed by a blank line. So I plan to abstain relative to whether this should be accepted.

I corrected the nit-pick, hope I didn't forget one.

henrikbrixandersen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,744 @@
+
+	/* No need to update the write offset for readonly */
+	if (dev_config->readonly) {
+		return 0;
```
Wont we miss filling in the optional RAM buffer on read-only devices?

Laczen replied:
Correct, I have corrected this by removing this early return. This will also allow a emulated eeprom that has changes written to flash to make a transition from rw to readonly.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,748 @@
+			break;
+		}
+
+		address += ((uint32_t)(*blk) << (8*i));
```
I think that the most common style is to separate operands form operators with one space.
I more places.

Laczen replied:
Yes, will correct this.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,748 @@
+		return -EIO;
+	}
+
+	if ((write_op) && (dev_config->readonly)) {
```
Maybe switch the `if (!device` with this one in places; this basically serves bad parameters to the function and there is no need for checking if device is ready if function parameters will fail further operation.

Laczen replied:
I will recheck the order and definitively move the return for len=0 down.

Laczen replied:
After reconsidering I have changed the order:
1. Check if there is nothing to do and return early
2. Do input error checking and return if error
3. Handle normal case

I will try to keep this in next pieces of code :-) :-).

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,748 @@
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	for (int i = 0; i < (dev_config->flash_cbs/2); i++) {
```
I think that `size_t` should be used instead of `int`.

Laczen replied:
Not really, `dev_config->flash_cbs` is a `uint8_t`.

pabigot replied:
As-is this is doing a signed comparison (because the division mixes unsigned and signed operands), so `int` is more appropriate than `size_t`.  For rigor it should be:

    for (uint8_t i = 0; i < (dev_config->flash_cbs/2U); i++) {

but honestly using `size_t` for the loop variable instead of `uint8_t` would not be a problem.  Matter of taste, though.

de-nordic commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,748 @@
+		(*blk++) = (*data++);
+	}
+
+	for (int i = 0; i < (dev_config->flash_cbs/2); i++) {
+		if (i < sizeof(address)) {
+			(*blk++) = (uint8_t)(((address >> (8 * i)) & 0xff));
+		} else {
+			(*blk++) = 0xff;
+		}
+
+	}
```
Am I missing something or this loop actually overwrites what the first loop has written to the buffer pointed by `blk`?

Laczen replied:
You are missing something, *blk goes from 0 to flash_cbs/2 - 1 in the first loop, and from flash_cbs/2 to flash_cbs - 1 in the second.

de-nordic reviewed (changes requested):
Left some comments.

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,634 @@
+	off_t addr_jmp;
+	int rc = 0;
+
+	if ((!data) || (!eeprom_emu_range_is_valid(dev, address, len))) {
+		return -EINVAL;
+	}
+
+	if (!len) {
+		return 0;
+	}
```
This seems a bit nit-picking :-). If a wrong address or no data is given, but the length is 0 it is considered OK. This length is really important overriding the wrong address or no data :-) :-) :-).

I always try to write routines as: do a check for input error first, return if wrong. Then handle the OK case and do nothing if you don't need to do anything.

Laczen commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,634 @@
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if ((address + len) <= dev_config->size) {
+		return 1;
+	}
+
+	return 0;
```
True, but can you always say that this will return 1, or should it be `return ((address + len) <= dev_config->size ? 1 : 0)`?

pabigot replied:
The value of a comparison operation implicitly or explicitly cast to an integer is 1.

Laczen committed:
drivers: eeprom: add emulated eeprom in flash

Modifications to incorporate latest write to new flash area
Modification to avoid writing garbage to new flash area when compactor
is called during init.
Modifications to allow erase at partition size instead of eeprom
pagesize.
Modifications to better separate rambuf usage from flash usage.
Corrected some errors in compactor

Signed-off-by: Laczen JMS <laczenjms@gmail.com>

de-nordic commented:
Does the emulation support non-0xff erasable flash? This needs to be tested.

Laczen commented:
> Does the emulation support non-0xff erasable flash? This needs to be tested.

No it does not support non-0xff eraseable flash.
It would be easy to add it but this only makes sense when all flash devices are reporting this, AFAIK not all flash devices are reporting this.
Also the devices that have a flash with erase value 0x00 (stm32l1 and stm32l0) have a hardware EEPROM so no need to use emulation in that case.

de-nordic reviewed (approved):
There are missing spaces within arithmetic operations but otherwise looks OK.

nvlsianpu commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,751 @@
+ * the size of the EEPROM (required address space), with a minimum of 4 byte.
+ * The EEPROM page needs to be a multiple of the flash page. Multiple EEPROM
+ * pages are also supported and increases the number of writes that can be
+ * performed.
```
would be nice to have here a sentence which describes how EEPROM pages usage are rotated.

nvlsianpu commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,751 @@
+			dev_data->write_offset += sizeof(buf);
+			if ((dev_data->write_offset + sizeof(buf)) >=
+			    dev_config->page_size) {
+				rc = eeprom_emu_compactor(dev);
```
It would be possible to detect GC need before any change record will be written.
Thanks to that refreshed `EEPROM data` entry will contain change at once.

Laczen replied:
It is not so easy to detect GC need,  a write can be done to a already used part and an unused part. However i do understand what you mean. I will give it some thinking and try to come up with a prposal (the proposal has been postponed to 2.6 anyhow).

Laczen replied:
The lastest commit should resolve this.

nvlsianpu commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,751 @@
+			return -EIO;
+		}
+
+		rc = eeprom_emu_compactor(dev);
```
:question:
`rambuf` (if available) is not initialized yet, In that case this call to `eeprom_emu_compactor()` will write a garbage to the EEPROM.
Might be I wrong?

Laczen replied:
You are right, I am going to rework `eeprom_emu_compactor()` so that it doesn't use `rambuf`.

nvlsianpu replied:
I think usage of valid `rambuf` is actually great idea, I appreciated that. Maybe you can just set the rambuf pointer to NULL for this call, and restore it once it return.

Laczen replied:
The latest commit should resolve this.

nvlsianpu reviewed (changes requested):
Implementations and idea looks fine to me.
Especially `rambuf` is great thing.

Need wait for answer to my question on possible bug in initialization routine.

nvlsianpu commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -323,15 +344,30 @@ static int eeprom_emu_compactor(const struct device *dev)
 
 	} else {
 		off_t rd_offset = 0;
+		uint8_t *data8 = (uint8_t *)(ctx->data);
```
`ctx` might be `NULL`

Laczen replied:
Correct, updated.

nvlsianpu commented:
I forgot about that flash protection should be disabled before each flash modify operation, see
https://docs.zephyrproject.org/latest/reference/peripherals/flash.html#c.flash_write_protection_set

Apart from above comment, all my previous comments were answered properly.

Laczen commented:
> I forgot about that flash protection should be disabled before each flash modify operation, see
> https://docs.zephyrproject.org/latest/reference/peripherals/flash.html#c.flash_write_protection_set
> 
> Apart from above comment, all my previous comments were answered properly.

@nvlsianpu, thanks for the comments and the review.

Regarding your last remark you forgot about: this has been taken into account: as any change to flash (either a write or an erase) will happen in the context of an ´eeprom_emu_write()´ operation and the flash protection disable/set is done in the ´eeprom_emu_write()´. This avoids having to change the flash protection for each write/erase call.

nvlsianpu commented on drivers/eeprom/eeprom_emulator.c:
```diff
@@ -0,0 +1,831 @@
+	if ((!dev_config->readonly) &&
+	    ((dev_data->write_offset + sizeof(buf)) >= dev_config->page_size)) {
+		/* Need to write in next step, disable flash write protection */
+		if (flash_write_protection_set(dev_config->flash_dev, false)) {
```
flash_write_protection_set(..., flase) in eeprom_emu_flash_write/erase are sufficient. No need to do it here.

Laczen replied:
Oops, forgot to remove in the init routine...
